pkg/pillar/types/disktypes.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2022 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package types
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+//EdgeNodeDiskDescription stores information to identify disk
+type EdgeNodeDiskDescription struct {
+	Name         string
+	LogicalName  string
+	SerialNumber string
+}
+
+//Equal return true if descriptions points onto the same disk
+func (dsc EdgeNodeDiskDescription) Equal(another EdgeNodeDiskDescription) bool {
+	if dsc.Name == another.Name && dsc.Name != "" ||
+		dsc.SerialNumber == another.SerialNumber && dsc.SerialNumber != "" ||
+		dsc.LogicalName == another.LogicalName && dsc.LogicalName != "" {
+		return true
+	}
+	return false
+}
+
+//MergeDescription append non-empty information from another description if points onto the same disk
+func (dsc EdgeNodeDiskDescription) MergeDescription(another EdgeNodeDiskDescription) EdgeNodeDiskDescription {
+	newDsc := dsc
+	if dsc.Equal(another) {
+		if another.Name != "" {
+			newDsc.Name = another.Name
+		}
+		if another.SerialNumber != "" {
+			newDsc.SerialNumber = another.SerialNumber
+		}
+		if another.LogicalName != "" {
+			newDsc.LogicalName = another.LogicalName
+		}
+	}
+	return newDsc
+}
+
+//Key for pubsub
+func (dsc EdgeNodeDiskDescription) Key() string {
+	if dsc.Name == "" && dsc.LogicalName == "" && dsc.SerialNumber == "" {
+		return ""
+	}
+	h := sha256.New()
+	h.Sum([]byte(dsc.Name))
+	h.Sum([]byte(dsc.LogicalName))
+	h.Sum([]byte(dsc.SerialNumber))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+//EdgeNodeDiskConfigType should be in sync with api
+type EdgeNodeDiskConfigType int32
+
+// enum should be in sync with api
+const (
+	EdgeNodeDiskConfigTypeUnspecified EdgeNodeDiskConfigType = iota // no configured, do nothing
+	EdgeNodeDiskConfigTypeEveOs                                     // the disk EVE is installed on
+	EdgeNodeDiskConfigTypePersist                                   // the disk is separate persist partition, not zfs
+	EdgeNodeDiskConfigTypeZfsOnline                                 // included in zfs and online
+	EdgeNodeDiskConfigTypeZfsOffline                                // included in zfs and offline
+	EdgeNodeDiskConfigTypeAppDirect                                 // for direct assignment
+	EdgeNodeDiskConfigTypeUnused                                    // removed from zfs/app-direct
+)
+
+//EdgeNodeDiskConfig disk configuration
+type EdgeNodeDiskConfig struct {
+	Disk    EdgeNodeDiskDescription
+	OldDisk *EdgeNodeDiskDescription
+	Config  EdgeNodeDiskConfigType
+}
+
+//Key is the key in pubsub
+func (cfg EdgeNodeDiskConfig) Key() string {
+	return cfg.Disk.Key()
+}
+
+//GetDiskNameByPartName returns disk name /dev/sda for part name /dev/sda1 if exists
+func GetDiskNameByPartName(name string) (string, error) {
+	link, err := os.Readlink(filepath.Join("/sys/class/block", filepath.Base(name)))
```

Comment:
You need to check for /sys/class/nvme too.

Reply:
Hmm, may be not. NVMe block devices will show up /sys/class/block too. So should be fine.